src/request.js

```diff
@@ -0,0 +1,237 @@
+        hostname: urlParts.hostname,
+        port: port || 443,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    case 'http:':
+      return makeRequest({
+        httpLib: http,
+        hostname: urlParts.hostname,
+        port: port || 80,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    default:
+      throw new RequestError('Invalid protocol: ' + urlParts.protocol);
+      break;
+  }
+  
+}
+
+module.exports = {
+  request,
+  RequestError,
+};
+
+async function makeRequest(opts) {
+  assert(typeof opts === 'object');
+  assert(typeof opts.httpLib === 'object');
+  assert(typeof opts.hostname === 'string');
+  assert(typeof opts.port === 'number');
+  assert(typeof opts.path === 'string');
+  assert(typeof opts.method === 'string');
+  assert(typeof opts.headers === 'object');
+  assert(typeof opts.timeout === 'number');
+  if (opts.stream) {
+    assert(typeof opts.contentType === 'string');
+    assert(typeof opts.stream === 'object');
+    assert(typeof opts.stream.pipe === 'function');
+  }
+  if (opts.data) {
+    assert(typeof opts.contentType === 'string');
+    assert(typeof opts.data.length === 'number');
+  }
+
+  return new Promise(async (resolve, reject) => {
+    let request = opts.httpLib.request({
+      hostname: opts.hostname,
+      port: opts.port,
+      path: opts.path,
+      method: opts.method,
+    });
+
+    // We don't want connections hanging around forever, so we'll set a fatal
+    // condition that they timeout if they aren't responded to quickly enough
+    request.setTimeout(opts.timeout);
+
+    request.on('error', err => {
+      reject(err);
+    });
+
+    request.on('socket', socket => {
+      debug('have a socket');
+
+      socket.on('error', err => {
+        debug('socket error');
+        reject(err);
+      });
+
+      socket.on('close', had_error => {
+        reject(new RequestError('Socket closed with error'));
+      });
+
+    });
+
+    request.on('response', response => {
+      if (!response.statusCode) {
+        reject(new RequestError('Request has invalid status code'));
+      }
+      response.socket.on('error', err => {
+        debug('response socket error: ' + err.stack || err);
+      });
+      resolve(response);
+    });
+
+    request.on('abort', () => {
+      reject(new RequestError('Request aborted by client (us)'));
+    });
+
+    request.on('aborted', () => {
+      reject(new RequestError('Request aborted by server (them)'));
+    });
+
+    request.on('timeout', () => {
+      request.abort();
+      reject(new RequestError('Request timeout occured'));
```

Comment:
No need to reject here... let the request be aborted.... The `abort` event will be emitted.

`request.on('timeout', () => request.abort());` should do the trick.